### DISCUSSION TITLE
Problem with DP-CTGAN parameters

### DISCUSSION THREAD

**Discussion by Dean-lyc** [opendp/opendp#517] [SmartNoise] [answered]
Hi Team,

I'm trying to sample data with DP-CTGAN. I followed your official document code:

`synth = PytorchDPSynthesizer(1.0, DPCTGAN())
synth.fit(pums, categorical_columns=list(pums.columns), transformer=BaseTransformer)`

When I set the first parameter in PytorchDPSynthesizer less than 1, I always get 'per_column_epsilon must be positive'. Is the first means epsilon in differential privacy? I think that can be any positive number.

When I set keep the first parameter to be 1.0,  no matter how many data I sample with `sample = synth.sample(10)` , I always get sample_rate=4.166666666666667 is not a valid value. Please provide a float between 0 and 1. I don't where can I change the sample_rate to fix this problem.

Best regards,
Yucong

**Comment by joshua-oss** [comment]
Tagging @AprilXiaoyanLiu

**Comment by joshua-oss** [comment, marked as answer]
I just realized what this is.  We added some code to transform continuous values, and added the `preprocessor_eps` parameter, which defaults to 1.  When budget for the synthesizer is set to 1.0, the preprocessor spends all of it, and there is some tiny floating point error left for the actual processing.  There is a check to throw an error if `preprocessor_eps` is more than the total budget available, but it's so close in this case that it isn't caught.

We are moving all preprocessing out of the synthesizers into a transforms class, so this bug should go away soon, but you can work around it in the meantime by setting `preprocessor_eps` to be very small, like:

```python
synth = PytorchDPSynthesizer(0.5, DPCTGAN(preprocessor_eps=0.01))
```

I wasn't able to reproduce the issue with `sample_rate` showing an error, but hopefully this fixes it.  Let me know if it still reproduces for you.
